TODOs.txt

```diff
@@ -0,0 +1,17 @@
+- Use a uniform API for all async Flight client APIs:
+
+  void Foo(const FlightCallOptions&, shared_ptr<ReadListener<T>> listener);
+
+  - shared_ptr, or raw pointer? (gRPC uses the latter...)
```

Comment:
Or weak_ptr? This way the caller can easily decide the lifetime (like a raw pointer), while not risking crashes (unlike a raw pointer).

Reply:
We could use non-const reference to express this more strongly? weak_ptr seems rather awkward to use here unless you already have a shared_ptr (in which case, you'd probably rather just pass shared_ptr), but I dislike forcing heap allocation here.

Reply:
The main challenge IMHO is to ensure proper lifetime management. If you don't pass a managed pointer (either shared or weak), the async machinery might invoke a listener that doesn't exist anymore.

By the way, some cancellation mechanism might also be desired?

Reply:
TryCancel is included as a method.

I can switch to forcing a shared pointer again...

Reply:
Hmm, the problem is then we have a circular reference (and weak_ptr just punts the problem). The caller has state (the callbacks), and the transport has state (the actual RPC), and they both need to reference each other (the callbacks need to update the transport, and the transport needs to invoke the callbacks). 

I suppose we could very carefully drop the weak/shared_ptr holding the callbacks from the RPC state when the RPC finishes, breaking the loop...? But I would rather not create the situation in the first place.

For what it's worth, the gRPC API expects the caller to hold the RPC state alive (that's the grpc::ClientContext).

I suppose a different solution would be to force the transport to hold all the state (e.g. by stashing it in the client object). This makes the client worse for concurrent RPCs (now they're all blocking on that central state, even if only for a short time).

Another might be to change the return type from `void` to some object containing the RPC state. The caller would then hold this alive (and internally it would own the callbacks). You would still have to heap-allocate the callback structure.

Reply:
I'll draft out the shared_ptr locally (I suppose you can always wrap a reference to a stack allocated object if the allocation matters)